pkg/kubernetes/client/v1alpha1/client.go

```diff
@@ -28,199 +28,252 @@ import (
+type getKubeClientsetFn func(*rest.Config) (*kubernetes.Clientset, error)
 
-// getKubernetesClientsetFunc provides the abstraction to get
-// kubernetes clientset
-//
-// NOTE:
-//  typed function makes it simple to mock
-type getKubernetesClientsetFunc func(*rest.Config) (*kubernetes.Clientset, error)
-
-// Client provides common kuberenetes client operations
+// Client provides Kubernetes client operations
 type Client struct {
-	IsInCluster    bool   // flag to let client point to its own cluster
-	KubeConfigPath string // kubeconfig path to get kubernetes clientset
+	// IsInCluster flags if this client points
+	// to its own cluster
+	IsInCluster bool
 
-	// Below functions are useful during mock
+	// KubeConfigPath to get kubernetes clientset
+	KubeConfigPath string
 
-	// handle to get in cluster config
-	getInClusterConfig getInClusterConfigFunc
+	// handle to get in-cluster config
+	getInClusterConfig getInClusterConfigFn
 
-	// handle to get desired kubernetes config
-	buildConfigFromFlags buildConfigFromFlagsFunc
+	// handle to get kubernetes config
+	// from flags
+	buildConfigFromFlags buildConfigFromFlagsFn
 
 	// handle to get kubernetes clienset
-	getKubernetesClientset getKubernetesClientsetFunc
+	getKubeClientset getKubeClientsetFn
 
-	// handle to get dynamic kubernetes clientset
-	getKubernetesDynamicClient getKubernetesDynamicClientFunc
+	// handle to get kubernetes dynamic clientset
+	getKubeDynamicClient getKubeDynamicClientFn
 
-	// handle to get kubernetes master IP
-	getKubeMasterIP getKubeMasterIPFunc
+	// handle to get kubernetes master IP from
+	// environment variable
+	getKubeMasterIPFromENV getKubeMasterIPFromENVFn
 
 	// handle to get kubernetes config path
-	getKubeConfigPath getKubeConfigPathFunc
+	// from environment variable
+	getKubeConfigPathFromENV getKubeConfigPathFromENVFn
 }
 
-// OptionFunc is a typed function that abstracts any kind of operation
-// against the provided client instance
+// OptionFn is a typed function to abstract
+// any operation against the provided client
+// instance
 //
-// This is the basic building block to create functional operations
-// against the client instance
-type OptionFunc func(*Client)
+// NOTE:
+//  This is the basic building block to create
+// functional operations against the client
+// instance
+type OptionFn func(*Client)
 
 // New returns a new instance of client
-func New(opts ...OptionFunc) *Client {
+func New(opts ...OptionFn) *Client {
 	c := &Client{}
 	for _, o := range opts {
 		o(c)
 	}
+
 	withDefaults(c)
 	return c
 }
 
 func withDefaults(c *Client) {
-	if c.getInClusterConfig == nil {
-		c.getInClusterConfig = rest.InClusterConfig
+	for _, def := range defaultFns {
+		def(c)
+	}
+}
+
+var defaultFns = []OptionFn{
+	withDefaultGetInClusterConfigFn(),
+	withDefaultBuildConfigFromFlagsFn(),
+	withDefaultGetKubeClientsetFn(),
+	withDefaultGetKubeDynamicClientFn(),
+	withDefaultGetKubeMasterIPFromENVFn(),
+	withDefaultGetKubeConfigPathFromENVFn(),
+}
```

Comment:
If we get in cluster config from `withDefaultGetInClusterConfigFn`  -- Do we still require to execute other `withs...`?

Reply:
No required.

However, this is a dumb list of settings that sets the default options.
Caller code need not bother about it at all.
There is no change in functionality when compared with older version with this version.
This version make the logic modular & also makes LOC in withDefaults less.